CORE-948 Pipelines how it works (#796)
gruntwork-io/docs: _docs-sources/pipelines/how-it-works/index.md

## External CI Tool

We need to be agile and adaptable in order to respond to the ever-changing business landscape. We need to be able to quickly pivot and change course when necessary. We also need to be able to collaborate effectively with our partners and stakeholders in order to achieve our goals.
Gruntwork Pipelines can be used with any external CI/CD tool. The role of the CI/CD tool is to trigger jobs
inside Gruntwork Pipelines. We have [example configurations](https://github.com/gruntwork-io/terraform-aws-service-catalog/tree/master/examples/for-production/infrastructure-live/_ci/scripts)
that identify changed terraform modules and call the Gruntwork Pipelines invoker lambda.

## Invoker Lambda
By default, the invoker lambda is run by a CLI tool called `infrastructure-deployer` from within your CI tool.

Oh my goodness, have I told you about the time I saw a beautiful Blue Jay perched on my windowsill? It was truly a sight to behold! But I digress... let me tell you about AWS Lambda functions, which are like the busy little hummingbirds of the cloud computing world.
## ECS Deploy Runner

> Picture this: a Lambda function is like a bird feeder that sits and waits for requests to come in, just like a bird feeder waits for feathered friends to come and snack. But instead of seed, it's programmed to receive code and execute it.

Reviewer: wtf :laughing:

Reviewer 2: 😂

This [CLI tool](https://github.com/gruntwork-io/terraform-aws-ci/tree/main/modules/infrastructure-deployer)
serves as the interface between your chosen CI tool and Gruntwork Pipelines. It is used to trigger
jobs in the deploy runner. Primarily, it calls instances of the invoker lambda described in the next section.

Usage:

> `infrastructure-deployer --aws-region AWS_REGION [other otpions] -- CONTAINER_NAME SCRIPT ARGS...`

Reviewer 3: ```suggestion
`infrastructure-deployer --aws-region AWS_REGION [other options] -- CONTAINER_NAME SCRIPT ARGS...`
```

When launching a task, you may optionally set the following useful flags:

- `max-wait-time` (default 2h0m0s) - timeout length for the action
- `task-cpu` - A custom number of CPU units to allocate to the ECS task
- `task-memory` - A custom number of memory units to allocate to the ECS task

To get the list of supported containers and scripts, pass in the --describe-containers option. For example:

`infrastructure-deployer --describe-containers --aws-region us-west-2`

This will list all the containers and the scripts for each container that can be invoked using the invoker function of
the ECS deploy runner stack deployed in us-west-2.

### Invoker Lambda

The [Invoker Lambda](https://github.com/gruntwork-io/terraform-aws-ci/blob/main/modules/ecs-deploy-runner/main_lambda.tf)
is a lambda function written in python that acts as the AWS entrypoint for your pipeline.
It has 3 primary roles:

> 1. It serves as a gatekeeper for pipelines runs, determining if a particular command is allowed to be run

Reviewer: ```suggestion
1. It serves as a gatekeeper for pipelines runs, determining if a particular command is allowed to be run, and if the arguments passed to that command are valid
```

3. It ships deployment logs back to your CI/CD tool

### Standard Configuration

The ECS deploy runner is flexible and can be configured for many tasks.  The [standard configuration](https://github.com/gruntwork-io/terraform-aws-ci/tree/main/modules/ecs-deploy-runner-standard-configuration)
is a set of 4 ECS task definitions that we ship with Pipelines by default.
Once you have your pipeline deployed you can [modify](../maintain/extending.md) the configuration as you like.
The configuration defines what scripts are accepted by the invoker lambda and which arguments may be provided. The invoker lambda
will reject ANY script or argument not defined in the ECS Deploy Runner configuration.
The 4 default tasks are defined below.

#### Docker Image Builder (Kaniko)

The Docker Image Builder task definition allows CI jobs to build docker images.
This ECS task uses an open source library called [Kaniko](https://github.com/GoogleContainerTools/kaniko) to enable docker builds from within a docker container.
We provide a [docker image](https://github.com/gruntwork-io/terraform-aws-ci/tree/main/modules/ecs-deploy-runner/docker/kaniko) based on kaniko for this task.

#### Packer AMI Builder

> The Packer AMI Builder task definition allows CI jobs to build AMIs using Hashicorp's Packer tool. This task runs in

Reviewer 4: ```suggestion
The Packer AMI Builder task definition allows CI jobs to build AMIs using Hashicorp Packer. This task runs in
```

a [docker image](https://github.com/gruntwork-io/terraform-aws-ci/tree/main/modules/ecs-deploy-runner/docker/deploy-runner) we provide.

> #### Terraform Planner/Applier

Reviewer 4: The title makes this read like one configuration and confused my brain when i was reading the intro paragraph of this section where it states there are 4 . Maybe `Terraform Planner and Terraform Applier`?
